internal/services/ingest.go

```diff
@@ -72,37 +80,64 @@ func NewIngestService(
 }
 
 func (m *ingestService) Run(ctx context.Context, startLedger uint32, endLedger uint32) error {
-	heartbeat := make(chan any)
-	go trackServiceHealth(heartbeat, m.appTracker)
+	rpcHeartbeat := make(chan entities.RPCGetHealthResult, 1)
+	ingestHeartbeat := make(chan any, 1)
+
+	// Start service health trackers
+	go trackRPCServiceHealth(ctx, rpcHeartbeat, m.appTracker, m.rpcService)
+	go trackIngestServiceHealth(ctx, ingestHeartbeat, m.appTracker)
+
 	if startLedger == 0 {
 		var err error
 		startLedger, err = m.models.Payments.GetLatestLedgerSynced(ctx, m.ledgerCursorName)
 		if err != nil {
 			return fmt.Errorf("erorr getting start ledger: %w", err)
 		}
 	}
+
 	ingestLedger := startLedger
-	for ; endLedger == 0 || ingestLedger <= endLedger; ingestLedger++ {
-		time.Sleep(10 * time.Second)
-		ledgerTransactions, err := m.GetLedgerTransactions(int64(ingestLedger))
-		if err != nil {
-			log.Error("getTransactions: %w", err)
-			continue
-		}
-		heartbeat <- true
-		err = m.ingestPayments(ctx, ledgerTransactions)
-		if err != nil {
-			return fmt.Errorf("error ingesting payments: %w", err)
-		}
+	for endLedger == 0 || ingestLedger <= endLedger {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context cancelled: %w", ctx.Err())
+		case resp := <-rpcHeartbeat:
+			switch {
+			// Case-1: wallet-backend is running behind rpc's oldest ledger. In this case, we start
+			// ingestion from rpc's oldest ledger.
+			case ingestLedger < resp.OldestLedger:
+				ingestLedger = resp.OldestLedger
+			// Case-2: rpc is running behind wallet-backend's latest synced ledger. We wait for rpc to
+			// catch back up to wallet-backend.
+			case ingestLedger > resp.LatestLedger:
+				log.Debugf("waiting for RPC to catchup to ledger %d (latest: %d)",
+					ingestLedger, resp.LatestLedger)
+				time.Sleep(5 * time.Second)
```

Comment:
Is this sleep of 5 seconds necessary? What happens if you remove it?

Reply:
@gouthamp-stellar I added the sleep here, so we give a little bit more time to rpc to catch up, instead of pinging it. Although, we can remove it since the goroutine invokes getHealth pings every 5s. What do you think?

Reply:
I dont see a negative of giving it extra time to catch up though